### ISSUE TITLE
Allow passing in a `format_spec` to Quantity.to_string

### ISSUE THREAD

**Issue by mhvk** [astropy/astropy#15933]
### What is the problem this feature will solve?

A left-over from #12571 by @misharash, where the bug part was fixed, but a more general comment was not, that  `Quantity.to_string(format='latex')` hardcodes that numbers below `1e7` use float notation, and above use exponential,
```
In [8]: (1e7*u.m).to_string(format='latex')
Out[8]: '$10000000 \\; \\mathrm{m}$'

In [9]: (1e8*u.m).to_string(format='latex')
Out[9]: '$1 \\times 10^{8} \\; \\mathrm{m}$'
```
This is because we use a default `format_spec` in 
https://github.com/astropy/astropy/blob/7759561c8d8760e657cee18c22636ed2ef9fd40a/astropy/units/quantity.py#L1472

It would be nice to be able to set this, maybe easiest by just passing in a `format_spec` so one can choose what to use.

Note that there is a similar issue for the generic format, though the change is only at `1e16`
```
In [18]: (1e15*u.m).to_string()
Out[18]: '1000000000000000.0 m'

In [19]: (1e16*u.m).to_string()
Out[19]: '1e+16 m'
```
This is done by numpy, via https://github.com/astropy/astropy/blob/7759561c8d8760e657cee18c22636ed2ef9fd40a/astropy/units/quantity.py#L1464-L1465

Here, one could put in something like `formatter={'all': lambda x: format(x, format_spec)}` (not tested!).

**Comment by MridulS** [comment]
Can't we just control this with the `precision` keyword here?

Changing the precision for `latex`
```
In [3]: (1e8*u.m).to_string(format='latex', precision=9)
Out[3]: '$100000000 \\; \\mathrm{m}$'
```

And with the following diff it seems to work with the general case too.
```diff 
-                    np.array2string(self.value, precision=precision, floatmode="fixed")
+                    np.array2string(self.value, precision=precision, floatmode="fixed",
+                                    formatter={'all': lambda x: f"{x:{format_spec}}"})
```

```python
In [4]: (1e16*u.m).to_string(precision=17)
Out[4]: '10000000000000000 m'
```

**Comment by mhvk** [comment]
@MridulS - part of the problem is that precision is used in two different ways here - to give the number of significant digits *and* decide when to give an exponent. E.g., I might want to always typeset something with an exponent, so that 123.45 gets typeset as `1.2345\times10^{2}` - that would be easy by giving a `format_spec`, but not using precision.

**Comment by Deen-dot** [comment]
![image](https://github.com/astropy/astropy/assets/80238871/f4dd7357-8188-4e08-a82a-5629f0f9f66d)

If I understand correctly, that's what you're looking for ^

It sounds good, it gives the user more flexibility in representing the quantity, I can see this also being used in other contexts, and it's also fairly simple to implement. I can send a pull request for my current implementation of it

**Comment by mhvk** [comment]
This was fixed by #16087 (though with a more general `formatter` argument)
